### PR TITLE
PWX-28882: Fix Nomad support issues for vSphere provider

### PR DESCRIPTION
**Pull request by nrevanna**
Configmap apis were used in vSphere provider which stopped px from
    coming up on Nomad. The following changes are made
        - Store interface moved to cloudops
        - DriveSetStore interface created with Store as composite. Contains
          interfaces with DriveSet datastructure.
        - GetLockTryDuration introduced
        - Renamed GetLockTimeout/SetLockTimeout to
          GetLockHoldTimeout/SetLockHoldTimeout.
        - Implemented IsKeyLocked() and LockWithKey() interfaces in kv_store
        - Removed noOpStore implementation. It was a stop gap solution for
          lack of implementation in kv_store

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** (optional)
Closes #

**Special notes for your reviewer**:

### CODE REVIEW

File: store/store_k8s.go
Comment (pureneelesh) [inline store/store_k8s.go:173]: can we safely assume that the value can always be converted to string without requiring any encoding e.g. base64? If yes, should we just change the interface to accept string instead of []byte to make that clear?

File: store/store_k8s.go
Comment (pureneelesh) [inline store/store_k8s.go:194]: I can't tell when we need to sanitize the key and when we don't. Also, not clear about when we need to lock. Please add comments to clarify. Also, functions could be named differently to distinguish if we are assuming that caller needs to have acquired a lock in some cases.

File: store/store_k8s.go
Comment (pureneelesh) [inline store/store_k8s.go:240]: do we actually see etcd errors from the k8s api server? I would have thought that the k8s API server will hide etcd errors from the clients.

File: store/store_kv.go
Comment (pureneelesh) [inline store/store_kv.go:48]: Name of the error seems misleading. It made me think that kvdb has not been setup yet. We could just use InvalidArgument or something like that?

File: vsphere/vsphere.go
Comment (pureneelesh) [inline vsphere/vsphere.go:87]: what happens if storeParams is nil? Should we return an error?

File: vsphere/vsphere.go
Comment (pureneelesh) [inline vsphere/vsphere.go:771]: should we call Unlock? I would not expect storeLock to have a valid value when lockErr is not nil. Perhaps just log a message?

File: store/store.go
Comment (nrevanna) [reply store/store.go:105]: I am doing that for NewKvStoreWithParams. 
configmap.New does that for NewK8sStoreWithParams

File: store/store.go
Comment (nrevanna) [reply store/store.go:118]: I'll change to return error if the name is empty.

Comment (nrevanna) [reply store/store.go:118]: I misread. I didn't want to add the customization with name and default timeout params in order to avoid some duplication. Let me revisit.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:135]: GetSanitizedK8sName was added as part of a different request.
The only time it will result in same sanitized key for two different keys is when the two keys differ in a special character.
Key@ and Key$.
At the time of implementation we determined that special characters shouldn't be a differentiator. The idea of mapping each special character to some Letter was also pitched but wasn't perceived to be necessary.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:174]: ref: PWX-27246
PutKey uses patchWithRetries which causes a race condition describe above.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:136]: We probably might have to do that at some point.
CreateKey is called with the hostname which could contain special characters which we had encountered and was fixed.
Today LockWithKey is mostly static names. Could possibly encounter the same issue in the future. I'll file a ticket to do that.

Comment (nrevanna) [reply store/store_k8s.go:136]: https://portworx.atlassian.net/browse/PWX-30053
I am looking to merge this today. I'll resolve all the comments I have. If I have time after that, I'll address this too. Just avoiding extra testing.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:194]: CreateKey and DeleteKey were only use in vSphere and vSphere hostname had special characters and that was the reason it was added. These are existing functions. If you have some specific suggestions I can do them. The larger effort I'd prefer to defer to a later time since I am on a time crunch.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:240]: I am not sure. May be @adityadani might have more context to this.

File: vsphere/vsphere.go
Comment (nrevanna) [reply vsphere/vsphere.go:87]: I'll probably do that and handle in the caller.

File: store/store_kv.go
Comment (nrevanna) [reply store/store_kv.go:48]: This is from the existing code and I feel this it's technically correct to call ErrKvdbNotInitialized since it's an indication that the bootstrap kvdb is not initialized.

File: store/store.go
Comment (nrevanna) [reply store/store.go:81]: I corrected the comment and now calling to Store name. It's the name for which there will be an entry in the bootstrap kv.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:173]: It appears so. May be @adityadani can add more context to this.

File: store/store_kv.go
Comment (adityadani) [inline store/store_kv.go:34]: Lets fail the NewKVStore operation if name is empty and force the callers to set a key, so that we won't end up causing key conflicts

File: store/store_kv.go
Comment (adityadani) [inline store/store_kv.go:88]: The previous code used to take a lock under `clouddrive/_lock/`. Should we follow the same suite to ensure locks are under a prefix? In this way we don't change the behavior for Lock() API

File: vsphere/vsphere.go
Comment (adityadani) [inline vsphere/vsphere.go:97]: Should this be 180*time.Second instead to make it same as the default kvdb lock timeout of 3 minutes?

**Review by adityadani — APPROVED** [review]
lgtm after comments are addressed

File: store/store.go
Comment (adityadani) [inline store/store.go:124]:
```suggestion
		// Two cases:
		// internal kvdb && kv is not nil
		// external kvdb
		if !internalKvdb {
		    if kvdb.Instance() == nil {// return kvdb is not initialized}
		    kv = kvdb.Instance()
		}
		if withParams {
			s, err = NewKVStoreWithParams(kv, name, lockTryDuration, lockHoldTimeout)
		} else {
			s, err = NewKVStore(kv, name)
		}
```

File: store/store_kv.go
Comment (adityadani) [reply store/store_kv.go:88]: I think thats what you are trying to do. Can we rename this variable to `lockPrefix` ?

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:92]: I thought it was not needed for any of the existing use cases. We can add when required. There might be some extra work for this in the kv_store side.

File: vsphere/vsphere.go
Comment (nrevanna) [reply vsphere/vsphere.go:97]: I'd probably reduce that to 100 seconds. The lock will kinda serialize node brinup, so I'd reduce it futher.

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:240]: changed

File: store/store_k8s.go
Comment (nrevanna) [reply store/store_k8s.go:173]: https://portworx.atlassian.net/browse/PWX-30066

**Pull request merged**
